lib/permission-api.js

```diff
@@ -0,0 +1,270 @@
+
+      // FIXME: I don't understand why, but removing the following setTimeout causes the subsequent
+      // setTimeout call (when resolving the promise) to hang on RN iOS. 
+      // This might be related to our general makeCallback issues.
+      setTimeout(() => {}, 1);
+
+      if (error) {
+        reject(error);
+      }
+      else {
+        try {
+          let syncedRealm = new _Realm(config);
+          user[specialPurposeRealmsKey][realmName] = syncedRealm;
+          //FIXME: RN hangs here. Remove when node's makeCallback alternative is implemented
+          setTimeout(() => resolve(syncedRealm), 1);
+        } catch (e) {
+          reject(e);
+        }
+      }
+    });
+  });
+}
+
+function createInManagementRealm(user, modelName, modelInitializer) {
+  return getSpecialPurposeRealm(user, '__management', managementSchema)
+    .then(managementRealm => {
+      return new Promise((resolve, reject) => {
+        try {
+          let o;
+
+          const listener = () => {
+            if (!o) {
+              return;
+            }
+
+            if (typeof o.statusCode === 'number') {
+              managementRealm.removeListener('change', listener);
+
+              if (o.statusCode === 0) {
+                setTimeout(() => resolve(o), 1);
+              }
+              else {
+                const e = new Error(o.statusMessage);
+                e.statusCode = o.statusCode;
+                e.manamgementObject = o;
+                setTimeout(() => reject(e), 1);
+              }
+            }
+          }
+
+          managementRealm.addListener('change', listener);
+
+          managementRealm.write(() => {
+            o = managementRealm.create(modelName, modelInitializer);
+          });
+        }
+        catch (e) {
+          reject(e);
+        }
+      });
+    });
+}
+
+module.exports = {
+  getGrantedPermissions(recipient) {
+    if (recipient && ['currentUser', 'otherUser', 'any'].indexOf(recipient) === -1) {
+      return Promise.reject(new Error(`'${recipient}' is not a valid recipient type. Must be 'any', 'currentUser' or 'otherUser'.`));
+    }
+
+    return getSpecialPurposeRealm(this, '__permission', permissionSchema)
+      .then(permissionRealm => {
+          let permissions = permissionRealm.objects('Permission')
+            .filtered('NOT path ENDSWITH "__permission" AND NOT path ENDSWITH "__management"');
+          
+          if (recipient === 'currentUser') {
+              permissions = permissions.filtered('userId = $0', this.identity);
+          }
+          else if (recipient === 'otherUser') {
+              permissions = permissions.filtered('userId != $0', this.identity);
+          }
+          return permissions;
+      });
+  },
+
+  applyPermissions(condition, realmUrl, accessLevel) {
+    if (!realmUrl) {
+      return Promise.reject(new Error('realmUrl must be specified'));
+    }
+
+    if (['none', 'read', 'write', 'admin'].indexOf(accessLevel) === -1) {
```

Comment:
Declare as constant `validAccessLevels` and do `validAccessLevels.join(',')` in error message?